winsup/cygwin/globals.cc

```diff
@@ -73,6 +81,7 @@ bool pipe_byte;
 bool reset_com;
 bool wincmdln = true;
 winsym_t allow_winsymlinks = WSYM_deepcopy;
+fakehard_t fake_hardlinks = FHARD_ondemand;
```

Comment:
Hmm, but `FHARD_ondemand` does not cause actual behavioral change in the case of NTFS (or any other scenario where hardlinks are supported).

The very reason I proposed this is that msys2 does not exactly work (as in, the installation will be more or less broken) when it is (managed) on filesystems that does not support hardlinks (unless you stay away from packages that have hardlinks -- but then IIRC even coreutils does, just that it won't be broken util there's an update), so I'm not sure what you are referring to with *"relying on the current behavior"*.

It merely avoids e.g. the need of managing an installation on NTFS and only use a copy of it on a FAT-family filesystem (i.e. fake hardlinks manually).

Reply:
Besides, if the behavior is not "socially problematic" in the case of symlink, I doubt that there can be sensical / strong enough reason for us to prefer behavior preservation over working and consistent behavior. (But of course, that purely my own stand. I'm sure there are a lot people out there would prefer things to stay broken and wrong as long as that that keeps their setups working.)

Reply:
So have you thought about an MSYS2 installation on an NTFS drive that is used on a FAT drive?

With my maintainer hat on, I really do not want to see such behavioral change, certainly when unintended consequences are not even considered fully.

Reply:
> So have you thought about an MSYS2 installation on an NTFS drive that is used on a FAT drive?

Uhm yes? As I said, that's what one can / should do until this is merged, and the copy will be have all hardlinks "dereferenced". And unless this is merged, one cannot exactly "manage" (i.e. update) the copy inplace, but will have to keep copying (well, or rsync).

Am I really the one missing something here or is it actually you?

Reply:
Correct me if I am wrong. Right now, trying to hardlink on FAT will trigger an EPERM. With your patch, it will not do that but quietly copy instead (unless specifically overridden), right up until the disk is full.

To you, this won't be an unexpected change of behavior. To everybody else, it will be.

Reply:
Or if this is about "people should be aware that have msys2 on FAT volume instead of NTFS volume will additionally require hardlinks to be fake", and *you* think we should rely on the EPERM for that (instead of having a note on some README or so), there's nothing for me to address either, since it's your personal preference (and I have no intention to persuade you or anyone on that).

And ironically, one will only realize it when he/she installs/updates a package with pacman. I doubt that he/she will have any problem extracting a new installation with any of the available package / approach.

Reply:
Look, I've been maintaining software for a long, long time, longer than Git exists. One develops a sense of what poses maintenance problems over the course of such a long time.

And changing the behavior in a way as you are trying to do is such a problem. Of course, it is not _your_ problem because you simply go away after this PR is merged. The onus is firmly on others' shoulders. In this instance, among the few others who maintain the MSYS2 runtime: me.

So I won't merge this PR as-is.

Having said that, if you change the default behavior back to the current behavior, I am inclined to merge it.

Reply:
As I said, I don't even care if this will be merged anymore, because at the the end of the day I realized I can't exactly get any advantage by having msys2 on (ex)FAT without paying some unreasonable price (the AU size problem).

The only reason I continued the discussion is that you sounded like you missed something and/or there's room for discussion.

Although my experience is not as old as git, I've seen enough maintainers / developers who are allergic to *any* change (even when it's logical / reasonable / actually a bug fix).

And you know what, that's totally alright, what's not is that some incline to make it sound like who proposed the change / fix is "not seeing the bigger picture" / "missing some of the caveats", without providing what exactly those can be (because he/she simply couldn't). And apparently, you are yet another one of them.

If all you wanted to say is "I just don't like it", please just say it at the very beginning instead of wasting both of our time.

Reply:
Btw, I have no intention to be credited for some decision that I can't even tell / am not even told why. I can't / won't stop you from making any further change you want and merge it though, if you have such desire.